### ISSUE TITLE
Remember last window position?

### ISSUE THREAD

**Issue by tmpm697** [labwc/labwc#1680]
It's kind of trivial in stacking window manager that you close a window and then re-open, it will appear the same position on screen like before.

This feature will really useful and improve user's experience.

Thanks.

**Comment by Consolatis** [comment]
Technically we could remember the last position, size and app_id on unmap, compare against our stored list on map, store the list in some file on exit and reload that file when starting up. Not saying that we should implement the feature but technically it would be doable. Maybe as a new placement strategy.

**Comment by tokyo4j** [comment]
FYI: [Draft: staging: Add ext-placement protocol for output-relative window placement preferences](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/247)

**Comment by tmpm697** [comment]
@ahesford For your use-case, it can simple like this: open same app with multiple windows --> the last window close position will be remembered.

**Comment by ahesford** [comment]
You've missed the point of my question about multiple windows. The issue is how to honor a saved position when multiple windows are open. Do you just open the first window in the saved position? Do you open them all on top of each other? Do you pick some other, equally arbitrary option? There are so many snags in any possible implementation that this will never quite feel right.

**Comment by tmpm697** [comment]
Yes, i understand that, if there's no window of that app --> restore the last used position.

When open multiple windows of the same app, new window just default below the last one (or the current active one) 10-20 pixels right and bottom.

Take this picture from how macos new window from same app behavior:
https://i.imgur.com/oN8D1j4.png

if out of space in screen, default to last position, is that solve all the case you mentioned?

**Comment by ahesford** [comment]
No, my position remains that this is behavior fraught with nuance that relies on information we do not and can never have, so I don't think we should expend the effort trying to hack out an approximation.

**Comment by tmpm697** [comment]
> No, my position remains that this is behavior fraught with nuance that relies on information we do not and can never have, so I don't think we should expend the effort trying to hack out an approximation.

I don't know why you think it's `behavior fraught`, as one of many users, I would like to have this feature and those behavior seem pretty straight forward to achieve. 

But probably we have to extend some upstream protocols in wayland that can make some hesitancy, is that all of your concerns? We do not have protocols to implement this feature and we don't have enough desire for this features from you as a dev?

**Comment by droc12345** [comment]
Apps keep track of their size (w/h) as well as placement (x/y).
In X, the apps ask the wm  to place things in a certain area, but that only worked part of the time.
With wayland it's much harder, because X and wayland do not look at things like global coordinates 
in the same way. 
X owns the whole screen and wm's can ask X to put things in a certain way, all within it's screen.
Wayland doesn't do that, nor is it likely too.

**Comment by tmpm697** [comment]
Well, I guess wayland can be a good example to extend those protocols.

How would be *security* concern for wayland to fix if that's just overlay the window to others ones?
Apps can't intervene others and window manager also not own the whole screen.

Sorry if I missed the concept, not actually start to work on it, just asking as a user here.

**Comment by droc12345** [comment]
If you're interested you can look at this: Add ext-placement protocol for window positioning in "zones" 
https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/264
and look at the comments, as well as looking at the preceding MR's to see what and why things are the way they are.

**Comment by tmpm697** [comment]
> If you're interested you can look at this: Add ext-placement protocol for window positioning in "zones" https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/264 and look at the comments, as well as looking at the preceding MR's to see what and why things are the way they are.

Thanks for the link, but it seems un-relatable to this issue, and it sound like very bad idea of "zones"

And why such all of those limits from that proposal?

For example for this issue, why don't have a `sandbox` for each apps with their own `handlers` completely separate from other apps/windows that they can *display* their windows freely in every position in a screen that controlled by compositor.

That way, compositor can give apps to do whatever they want in their own environments, and I believe this is the way macos/windows manage their windows.

I can sense that proposal wants to minimize resources but is that all? Reduce some resources but more complexity over good diagrams and better code?

**Comment by tmpm697** [comment]
> Yes, i understand that, if there's no window of that app --> restore the last used position.
> 
> When open multiple windows of the same app, new window just default below the last one (or the current active one) of that app 10-20 pixels to the right and bottom.
> 
> Take this picture from how macos new window from same app behavior: https://i.imgur.com/oN8D1j4.png
> 
> if out of space in screen, default to last position, is that solve all the cases you mentioned?

And as you can see in the picture, we don't need a *zone* to do that, just overlay window and make it lower then the current active one.

**Comment by johanmalm** [comment]
@tmpm697  Welcome to the community

I'm not sure how you are receiving the feedback, so this message might not be needed, but just a couple of thoughts from me as lead maintainer to help manage expectations, feelings and behvaviors:

1. Take the feedback from core devs/maintainers for what it is - voices of experience based on knowledge of the wider eco-system, in-depth understanding of how the design hangs together as well as a good sense of how to keep a project maintainable, and - in response to feature/change requests - voices that speaks with intuition and that are (in my experience) often right, but of course sometimes wrong :smile:

2. With this being FOSS we only decide on what we merge - not what people work on! When people go off and work on things that we initially pushed back on, but that turns out to make sense, we merge them (and vice versa). What can be difficult is if people invest significant time in thinking through a proposal as well as designing/implementing it - and it still doesn't hit the mark and therefore isn't merged - but where the author was convinced all along that it would be merged. Luckily we haven't had that situation on this projects yet (that I'm aware of anyway).

3. In addition to the engineering, this is also about priorities and time-management. Unless other devs/maintainers respond on technical matters, I try to within a few days, but there is a lot going on, so I may not be able to engage in ongoing deep discussions on every issue. Where I'm going with this is that after a few initial exchanges of 'intuition' and feelings, if someone wants to pursue a particular line and explore something they feel is worth investing time in, the best next stage is generally to outline a well thought through technical evaluation which is a lot easier for me/others to respond to.

4. When someone new joins the community it can be hard to know how to pitch a conversation. Spelling out intent early can really help here. For example, do they have the means and ability to join in and do some of the work, or are 'feature requests' merely that (which is of course a valid position and absolutely fine)? Will they hang around for a long time and help fix/maintain things when they don't work. Are they committed to dedicating significant time to see a patch set through until point of (hopefully) merge. A good way to test this and build confidence (both ways) is to pick some simpler issues first before tackling harder and more controversial ones.

Hopefully the above is helpful.

Technically I don't get the proposal, so am not sure how to respond to this one... but just some quick off the top of ones head thoughts:
- Would it apply to all toplevels windows, for example modal dialogs? Or would it only apply to _root_ toplevel windows (i.e. those which do not have child toplevel surfaces).
- Considering that xdg-shell clients do not emit requests for x/y position, placement falls to the compositor with no guidance from the client. So what should the logic be with respect to app_id, PID, title, and so on? For example, should the window position only be remembered when the last open root-toplevel with a particularly app_id is closed?
- What should happen if the output-layout changes, or for example the scale of an output?
- What is the state of the protocol landscape for this?
- What do other compositors do? Have you checked COSMIC?
